### PR TITLE
Laurat007 2

### PR DESCRIPTION
**Pull request by laurat007**
This program solves both Q1 and Q2. I am not sure if the way I build my binary tree and the way I test the program are right.

### CODE REVIEW

File: laurat007/assignment2/Ancestors.java [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+        
+        // add to balance tree
+        // similar to bfs traversal
+        void addBalanced(T newValue, int skips) { 
+            BinaryTreeNode aux = this;
+            BinaryTreeNode newNode = new BinaryTreeNode(newValue);
+            Queue<BinaryTreeNode> queue = new LinkedList();
+            queue.add(aux);
+            
+            while(true) {
```
Comment (kkasiacodeu) [inline laurat007/assignment2/Ancestors.java]: Maybe better make it while(!queue.isEmpty()) instead of while(true). It's never good to see while(true) in code... Unless you are writing some service that is supposed to loop indefinitely

File: laurat007/assignment2/Ancestors.java [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+                aux = aux.right;
+            }
+            BinaryTreeNode newNode = new BinaryTreeNode(newValue);
+            aux.right = newNode;
+        }
+        
+        // add to balance tree
+        // similar to bfs traversal
+        void addBalanced(T newValue, int skips) { 
+            BinaryTreeNode aux = this;
```
Comment (kkasiacodeu) [inline laurat007/assignment2/Ancestors.java]: You can remove the aux variable altogether and just call q.add(this) here.

File: laurat007/assignment2/Ancestors.java [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+                    }
+                }
+            }
+        }
+        
+        public String toString() {
+            String s = "";
+            BinaryTreeNode aux = this;
+            Queue<BinaryTreeNode> q = new LinkedList();
+            q.add(aux);
```
Comment (kkasiacodeu) [inline laurat007/assignment2/Ancestors.java]: As above: you can remove the aux variable here.

File: laurat007/assignment2/Ancestors.java (hunk 1 above)
Comment (laurat007) [reply laurat007/assignment2/Ancestors.java]: Done. Thank you for advice! I found it really useful.

File: laurat007/assignment2/Ancestors.java (hunk 2 above)
Comment (laurat007) [reply laurat007/assignment2/Ancestors.java]: Done

File: laurat007/assignment2/Ancestors.java (hunk 3 above)
Comment (laurat007) [reply laurat007/assignment2/Ancestors.java]: Done

File: laurat007/assignment2/AncestorsTest.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+import java.util.ArrayList;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import junit.framework.TestCase;
+import org.junit.After;
+import org.junit.AfterClass;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import static org.junit.Assert.*;
```
Comment (kkasiacodeu) [inline laurat007/assignment2/AncestorsTest.java]: Try to avoid importing the whole package. This can lead to some double imports sometimes. It is usually better to import only what you are using from a package :)

File: laurat007/assignment2/AncestorsTest.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        tree.addBalanced(6, 1);
+        return tree;
+    }
+    
+    @BeforeClass
+    public static void setUpClass() {
+    }
+    
+    @AfterClass
+    public static void tearDownClass() {
```
Comment (kkasiacodeu) [inline laurat007/assignment2/AncestorsTest.java]: I don't think you actually need those two @BeforeClass and @AfterClass methods at all :) They will be triggered and run but since you don't do anything in them you shouldn't even define them

File: laurat007/assignment2/Ancestors.java [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+ * but if the value was not found yet, we test if it was found in the current
+ * subtree (in either left or right subtree of the current root) - if this is 
+ * true we add the root to solution and return true, else return false.
+ * 
+ *  The method solving the second task uses the first method as follows:
+ * there are two values given; for each one it finds its ancestors and it 
+ * searches the first common value.
+ * 
+ */
+class AncestorsTest {
```
Comment (kkasiacodeu) [inline laurat007/assignment2/Ancestors.java]: Remove this class from Ancestors.java

File: laurat007/assignment2/AncestorsTest.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import static org.junit.Assert.*;
+
+/**
+ *
+ * @author Laura
+ */
+public class AncestorsTest extends TestCase{
```
Comment (kkasiacodeu) [inline laurat007/assignment2/AncestorsTest.java]: Generally looks good, well done :)

Can you maybe try to add a couple more tests? For getAncestors method and for common ancestor. Try to have at least 2 or 3 tests for each, and maybe test edge cases like when there is no node with a given value etc

File: laurat007/assignment2/AncestorsTest.java (hunk 5 above)
Comment (laurat007) [reply laurat007/assignment2/AncestorsTest.java]: done.

Comment (laurat007) [reply laurat007/assignment2/AncestorsTest.java]: They were already there. It was not clear for me if they were supposed to be modified.

File: laurat007/assignment2/Ancestors.java (hunk 6 above)
Comment (laurat007) [reply laurat007/assignment2/Ancestors.java]: done.

File: laurat007/assignment2/AncestorsTest.java (hunk 7 above)
Comment (laurat007) [reply laurat007/assignment2/AncestorsTest.java]: added more tests.

File: laurat007/assignment2/AncestorsTest.java (hunk 4 above)
Comment (laurat007) [reply laurat007/assignment2/AncestorsTest.java]: It seems it was not even necessary that import.

File: laurat007/assignment2/AncestorsTest.java [hunk 8]
```diff
@@ -14,17 +14,16 @@
...
 
 /**
  *
  * @author Laura
  */
-public class AncestorsTest extends TestCase{
-    
+public class AncestorsTest extends TestCase {
+
     public AncestorsTest() {
```
Comment (kkasiacodeu) [inline laurat007/assignment2/AncestorsTest.java]: You don't need an empty constructor, java generates it if you haven't defined your own custom constructor.

File: laurat007/assignment2/AncestorsTest.java [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+        try {
+            super.tearDown();
+        } catch (Exception ex) {
+            Logger.getLogger(AncestorsTest.class.getName()).log(Level.SEVERE, null, ex);
+        }
+    }
+
+    // test getAncestors
+    @Test
+    public void Test1() {
```
Comment (kkasiacodeu) [inline laurat007/assignment2/AncestorsTest.java]: Try naming your tests in a more descriptive way. There are many conventions, for example you can have the name tell what it tests, what are special conditions and the expected outcome, for example: getAncestors_invalidNodes_returnsNull() etc.
Reading this method name you immediately can figure out what it tests and if you see this test failing you have an easy way to figure out what you might have broken with your changes. It benefits everyone :)

File: laurat007/assignment2/AncestorsTest.java (hunk 8 above)
Comment (laurat007) [reply laurat007/assignment2/AncestorsTest.java]: Removed constructor

File: laurat007/assignment2/AncestorsTest.java (hunk 9 above)
Comment (laurat007) [reply laurat007/assignment2/AncestorsTest.java]: Hope it's better now.
